pkg/collector/corechecks/oracle/activity.go

```diff
@@ -0,0 +1,211 @@
+package oracle
+
+import (
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/oracle/common"
+	"github.com/DataDog/datadog-agent/pkg/obfuscate"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// ActivitySnapshot is a payload containing database activity samples. It is parsed from the intake payload.
+// easyjson:json
+type ActivitySnapshot struct {
+	Metadata
+	// Tags should be part of the common Metadata struct but because Activity payloads use a string array
+	// and samples use a comma-delimited list of tags in a single string, both flavors have to be handled differently
+	Tags               []string            `json:"ddtags,omitempty"`
+	CollectionInterval float64             `json:"collection_interval,omitempty"`
+	OracleActivityRows []OracleActivityRow `json:"oracle_activity,omitempty"`
+}
+
+const ACTIVITY_QUERY = `SELECT /* DD_ACTIVITY_SAMPLING */
+	sid,
+	serial#,
+	username,
+    osuser,
+    process, 
+    machine, 
+    program ,
+    type,
+    sql_id,
+	force_matching_signature,
+    sql_plan_hash_value,
+    sql_exec_start,
+    module,
+    action,
+    client_info,
+    logon_time,
+    client_identifier,
+    CASE WHEN blocking_session_status = 'VALID' THEN
+	  blocking_instance
+	ELSE
+	  null
+	END blocking_instance,
+    CASE WHEN blocking_session_status = 'VALID' THEN
+		blocking_session
+	ELSE
+		null
+	END blocking_session,
+	CASE WHEN final_blocking_session_status = 'VALID' THEN
+    	final_blocking_instance
+	ELSE
+		null
+	END final_blocking_instance,
+	CASE WHEN final_blocking_session_status = 'VALID' THEN
+    	final_blocking_session
+	ELSE
+		null
+	END final_blocking_session,
+    CASE WHEN state = 'WAITING' THEN
+		event
+	ELSE
+		'CPU'
+	END event,
+	CASE WHEN state = 'WAITING' THEN
+    	wait_class
+	ELSE
+		'CPU'
+	END wait_class,
+	sql_text,
+	pdb_name
+FROM sys.dd_session
+WHERE 
+	( sql_text NOT LIKE '%DD_ACTIVITY_SAMPLING%' OR sql_text is NULL ) 
+	AND (
+		NOT (state = 'WAITING' AND wait_class = 'Idle')
+		OR state = 'WAITING' AND event = 'fbar timer' AND type = 'USER'
+	)`
+
+type RowMetadata struct {
+	Commands       []string `json:"dd_commands,omitempty"`
+	Tables         []string `json:"dd_tables,omitempty"`
+	Comments       []string `json:"dd_comments,omitempty"`
+	QueryTruncated string   `json:"query_truncated,omitempty"`
+}
+
+type OracleActivityRow struct {
+	SessionID              uint64  `db:"SID" json:"sid,omitempty"`
+	SessionSerial          uint64  `db:"SERIAL#" json:"serial,omitempty"`
+	Username               *string `db:"USERNAME" json:"username,omitempty"`
```

Comment:
Is there a particular reason for the pointers on the string fields? The zero value (empty string) would result in a field not being emitted in the json payload so unless we need to be able to distinguish between absence of a field and the value being empty, it would be nicer to use to have those set as `string`.

Reply:
I used pointers because `null` translates to `nil`, and, therefore can't be assigned to `string` or `int`. First, I tried to use `sql.NullString` instead, but `sql.NullString` is a structure that would require additional processing, so I switched to pointers. We can go away from pointers if we discover cases where `0` or `""` semantically wouldn't be the same as `null`.

Reply:
👍 , let's go with that and readjust if we need.